memory_w25q1128jv_spi/memory_w25q1128jv_spi.h

```diff
@@ -0,0 +1,186 @@
+/*
+ * W25Q128JV_MEMORY.h
+ *
+ *  Created on: Jan 1, 2024
+ *      Author: rober
+ */
+#pragma once
+
+#include <cmath>
+
+#if defined(STM32F1)
+#include "stm32f1xx_hal.h"
+#endif
+
+#ifndef SPI_Timeout
+#define SPI_Timeout 10
+#endif
+
+class MemoryW25q1128jvSpi {
+  public:
+    enum class State { CHIP_ENABLE, PAGE_WRITE, COMPLETE, ERROR };
+
+    /**
+     * @param hspi 		 SPI bus handler
+     * @param csPort 	 chip select GPIO port
+     * @param csPin 	 chip select GPIO pin
+     * @param holdPort  Write HOLD GPIO port
+     * @param holdPin 	 Write HOLD GPIO pin
+     * @param wpPort 	 Write Protect GPIO port
+     * @param wpPin	 Write Protect GPIO pin
+     */
+    MemoryW25q1128jvSpi(SPI_HandleTypeDef *hspi, GPIO_TypeDef *csPort, uint16_t csPin, GPIO_TypeDef *holdPort, uint16_t holdPin, GPIO_TypeDef *wpPort, uint16_t wpPin);
+
+    /**
+     * @brief Initialises the memory by finding the first free 64 bytes
+     * @retval Operation Sucess, 0 for success
+     * @retval Operation Failure, -1 for wrong devoce
+     */
+    int Init();
+
+    /**
+     * @brief Reads the device ID of the memory/
+     * 		  Should output 0xEF for manufacturer ID,
+     * 		  Should output 0x90 for Device ID
+     * @retval Operation Success, 0xEF17 is return value
+     * @retval Operation Failure, -1 for SPI protocol failure
+     */
+    int DeviceID();
+
+    /**
+     * @brief Reads the first status register from memort
+     * @retval Operation Sucess, Register value from 0-15 for success
+     * @retval Operation Failure, -1 for SPI protocol failure
+     */
+    int ReadStatusReg1();
+
+    /**
+     * @brief Erases all data on chip
+     * @retval Operation Sucess, 0 for success
+     * @retval Operation Failure, -1 for SPI protocol failure
+     */
+    int ChipErase();
+
+    /**
+     * @brief Writes a 64 byte array into flasg memory,
+     *        Auto increments from address found in Init
+     * @retval Operation FSM status
+     */
+    State ChipWrite(uint8_t (&data)[64]);
+
+    /**
+     * @brief Reads 64 bytes starting from read_address input,
+     *        ChipRead outout value is updated in input chipData
+     * @retval Operation Sucess, 0 for success
+     * @retval Operation Failure, -1 for SPI protocol failure
+     */
+    int ChipRead(uint32_t read_address, uint8_t (&chipData)[64]);
+
+    /**
+     * @brief Dumps each 64 byte data package from address 0,
+     *        auto increments each 64 byte data package
+     * @retval Operation Sucess, current read address value for success
+     * @retval Operation Failure, -1 for SPI protocol failure
+     */
+    int ChipReadDump(uint8_t (&chipData)[64]);
+
+    /* 64 byte structs for use in UCIRP projects*/
+#pragma pack(push, 1)
```

Comment:
the driver doesn't need to / should not care what packet its storing

Reply:
Agreed, none of the function in my code use these packets.

I left the packets in here so that whenever someone uses our memory module, the packets come imported with the memory drivers. I can take it out if you want, but I like the idea that other people who use the memory drivers can simply just define their needed packet without needing to define the struct in their main.